### DISCUSSION TITLE
Can an element be designed so it behaves differently in a HW vs. an Exam?

### DISCUSSION THREAD

**Discussion by firasm** [PrairieLearn/PrairieLearn#6108] [Q&A] [answered]
A general question: how can an element be designed so it behaves differently in a HW vs. an Exam?

The use-case for this is that I have a question that I've added some hints to, but when I use that same question in an exam, I'd like to disable the hints.

I don't want to have multiple copies of the same question and I know about the `<pl-hide-in-panel>` element, I do a mass find and replace before the exam to disable hints, but I'm wondering if there's a better way?

For instance: 

- by creating a new element that behaves differently when the assessment it's in is a HW vs. an Exam?
- Adding a flag in the assessment to enable/disable hints, similar to `showCorrectAnswer: false` that was recently added.

**Comment by echuber2** [comment]
Maybe someone else has a more elegant solution, but if you're willing to have a build step, you could use some templating or markdown format to specify the question and generate both the exam and non-exam versions with a script. This does result in two copies of the same question, but it may help to manage revisions to the question.

In general, I think questions don't have awareness of the context of the assessment they're in. However you could also remove the hints and present them to the student up-front in a separate dummy question on the HW only, and just omit that preface on the exam.

**Reply by echuber2** [reply]
You could propose a feature `<pl-hide-in-exam>` that would do what you expect, but I'm not sure how easy this will be to implement.

**Reply by firasm** [reply]
Yea, I have a markdown system I use to generate my problems, so I could do this pretty easily with a build steps that'd create a different version of my questions without hints.

so the path to the question would be nearly identical, I'd just replace `hw` with `exam`.

Neat idea, thanks!

**Comment by eliotwrobson** [comment]
I'm not 1000% sure on the differences between homework and exam mode, but we have an element in our course that gives progressive hints based on the number of incorrect submissions so far. So in an exam, if only a single submission is allowed, then the hints won't ever be displayed.

From a development standpoint, I think the easiest thing to do is to add an `assessment_type` field in the data dictionary that gets passed to all of the elements, and then write a custom hints element that has the behavior you want (which should be pretty simple).

**Reply by echuber2** [reply]
How did you implement the branching based on the number of incorrect submissions?

**Reply by firasm** [reply]
thanks @eliotwrobson ! I like the idea of giving progressive hints based on number of incorrect submissions. Does this mean that you put the hints inside one of the panels (submitted answer, correct answer, etc)?

**Reply by eliotwrobson** [reply]
We actually put the hints in the question panel with a custom element which tracks the number of valid submissions (for an appropriate definition of valid). Each time that the elements' grade function gets executed (and there are no format errors), it increments a counter local to the element. I'd be happy to share the code for this with you @firasm if you'd like to play around with it.

**Reply by mwest1066** [reply]
@eliotwrobson are you storing the local counter in `params`?

I seem to remember someone (maybe @ffund?) wanting to pass all previous submissions to to question functions, but I couldn’t quickly find an issue tracking that. I guess that would make it simpler to do your sort of hint display?

**Reply by eliotwrobson** [reply]
@mwest1066 yes, it's stored in params

**Reply by mwest1066** [reply]
Cool, makes sense. We should certainly consider generic mechanisms for making this information available (without needing to store counters, which is of course fine in itself).

**Reply by eliotwrobson** [reply]
Yes, would be cleaner if the data dict natively held some information about previous number of submissions, or something like that.

**Reply by firasm** [reply]
thanks @eliotwrobson !! I think the trick that I couldn't see before was storing/updating the counter in `params`. That's a really neat idea :-) and I see a path forward in the interim.

**Reply by ffund** [reply]
Does this only work if the student gets multiple attempts on the same variant? If a student gets a new variant on each attempt, then `params` will be all new, right?

**Reply by mwest1066** [reply]
That's correct, this only works on the same variant.

**Comment by nwalters512** [comment, marked as answer]
I think question parameters (https://github.com/PrairieLearn/PrairieLearn/pull/3848) will be the canonical way to do this once that PR lands. Unfortunately I don't have any great solutions that you can use immediately!

In the long run, we'd like to eliminate the notion of Homework/Exam assessment types entirely, so I would like to avoid exposing that concept to questions or elements.

**Reply by firasm** [reply]
yes!! Skimmed the stuff in the PR, and Question parameters is exactly what I want. I want to be able to control (some parts of) how the question behaves at the assessment level. Thank you @nealterrell for your work in #3848 🙏🏽.

Also, totally agree with eliminating assessment types, it should all be configurable settings at the assessment level.

**Reply by mwest1066** [reply]
@firasm please feel free to hassle us to make sure that #3848 gets landed. @nwalters512 and I need to make a final decision there, I think.
